compiler_opt/es/blackbox_optimizers.py

```diff
@@ -263,7 +277,141 @@ def run_step(self, perturbations: npt.NDArray[np.float32],
     # gradients
     return self.gradient_ascent_optimizer.run_step(current_input, gradient)
 
-  def get_hyperparameters(self):
+  def get_hyperparameters(self) -> List[float]:
+    if self.hyperparameters_update_method == UpdateMethod.STATE_NORMALIZATION:
+      return self.mean_state_vector + self.std_state_vector
+    else:
+      return []
+
+  def get_state(self) -> List[float]:
+    gradient_ascent_state = self.gradient_ascent_optimizer.get_state()
+    if self.hyperparameters_update_method == UpdateMethod.STATE_NORMALIZATION:
+      current_state = [self.nb_steps]
+      current_state += self.sum_state_vector
+      current_state += self.squares_state_vector
+      current_state += self.mean_state_vector + self.std_state_vector
+      return current_state + gradient_ascent_state
+    else:
+      return gradient_ascent_state
+
+  def update_state(self, evaluation_stats: SequenceOfFloats) -> None:
+    if self.hyperparameters_update_method == UpdateMethod.STATE_NORMALIZATION:
+      self.nb_steps += evaluation_stats[0]
+      evaluation_stats = evaluation_stats[1:]
+      first_half = evaluation_stats[:self.state_dim]
+      second_half = evaluation_stats[self.state_dim:]
+      self.sum_state_vector = [
+          sum(x) for x in zip(self.sum_state_vector, first_half)
+      ]
+      self.squares_state_vector = [
+          sum(x) for x in zip(self.squares_state_vector, second_half)
+      ]
+      self.mean_state_vector = [
+          x / float(self.nb_steps) for x in self.sum_state_vector
+      ]
+      mean_squares_state_vector = [
+          x / float(self.nb_steps) for x in self.squares_state_vector
+      ]
+
+      self.std_state_vector = [
+          math.sqrt(max(a - b * b, 0.0))
+          for a, b in zip(mean_squares_state_vector, self.mean_state_vector)
+      ]
+
+  def set_state(self, state: SequenceOfFloats) -> None:
```

Comment:
ah, so these APIs on MonteCarlo weren't used before, correct?

do they have test?

Reply:
They were here before and are present in the current codebase, but oddly the diff misrepresents it as new even after rebasing.